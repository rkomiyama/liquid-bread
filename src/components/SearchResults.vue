<template>
  <v-data-table
    :items="results"
    :headers="headers ? headers : []"
    :loading="loading"
    :search="search"
    :rows-per-page-items="rowsPerPageItems"
    :expand="expand"
    :custom-sort="customSort"
  >
    <v-progress-linear v-slot:progress color="blue" indeterminate></v-progress-linear>
    <template v-slot:items="props">
      <tr @click="props.expanded = !props.expanded">
        <td v-for="(column, i) in headers" :key="i" class="text-xs-left">
          <span
            v-if="column.value === 'volume' || column.value === 'boil_volume'"
          >{{ props.item[column['value']]['value'] }} {{ props.item[column['value']]['unit'] }}</span>
          <span v-else>{{ props.item[column['value']] }}</span>
        </td>
      </tr>
    </template>
    <template v-slot:expand="props">
      <BeerCard :beer="props.item"/>
    </template>
  </v-data-table>
</template>

<script>
import BeerCard from "./BeerCard/BeerCard";

export default {
  name: "SearchResults",
  methods: {
    customSort(items, index, isDesc) {
      items.sort((a, b) => {
        if (index === "first_brewed") {
          if (!isDesc) {
            return this.compareDates(a.first_brewed, b.first_brewed);
          } else {
            return this.compareDates(b.first_brewed, a.first_brewed);
          }
        } else {
          if (!isDesc) {
            return a[index] < b[index] ? -1 : 1;
          } else {
            return b[index] < a[index] ? -1 : 1;
          }
        }
      });
      return items;
    },
    compareDates(a, b) {
      const yearA = a.slice(-4);
      const yearB = b.slice(-4);
      if (yearA !== yearB) {
        return yearA < yearB ? -1 : 1;
      } else {
        const monthA = a.slice(0, 2);
        const monthB = b.slice(0, 2);
        return monthA < monthB ? -1 : 1;
      }
    }
  },
  components: {
    BeerCard
  },
  data() {
    return {
      expand: false
    };
  },
  props: {
    headers: {
      type: Array,
      default: () => []
    },
    loading: {
      type: Boolean,
      default: false
    },
    results: {
      type: Array,
      default: () => []
    },
    search: {
      type: String,
      default: ""
    },
    rowsPerPageItems: {
      type: Array,
      default: () => []
    }
  }
};
</script>

<style lang="scss" scoped>
</style>
