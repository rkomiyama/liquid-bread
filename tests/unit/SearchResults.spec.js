import Vue from "vue";
import Vuetify from "vuetify";
import { shallowMount } from "@vue/test-utils";
import { createRenderer } from "vue-server-renderer";
import SearchResults from "@/components/SearchResults.vue";
import searchHeaders from "#/mockData/searchColumnHeaders.json";
import searchResults from "#/mockData/searchResults.json";

Vue.use(Vuetify);

let wrapper;

describe("SearchResults.vue", () => {
  beforeEach(() => {
    wrapper = shallowMount(SearchResults, {
      propsData: {
        headers: searchHeaders,
        results: searchResults
      }
    });
  });

  it("sorts correctly using customSort", () => {
    const sortedByDescDate = wrapper.vm.customSort(
      searchResults,
      "first_brewed",
      true
    );
    sortedByDescDate.reduce(function(accumulator, currentValue) {
      expect(
        wrapper.vm.compareDates(
          accumulator.first_brewed,
          currentValue.first_brewed
        )
      ).toEqual(1);
      return currentValue;
    });

    const sortedByAscDate = wrapper.vm.customSort(
      searchResults,
      "first_brewed",
      false
    );
    sortedByAscDate.reduce(function(accumulator, currentValue) {
      expect(
        wrapper.vm.compareDates(
          accumulator.first_brewed,
          currentValue.first_brewed
        )
      ).toEqual(-1);
      return currentValue;
    });

    const sortedByIbuDesc = wrapper.vm.customSort(searchResults, "ibu", true);
    sortedByIbuDesc.reduce(function(accumulator, currentValue) {
      expect(accumulator.ibu > currentValue.ibu).toBeTruthy();
      return currentValue;
    });

    const sortedByIbuAsc = wrapper.vm.customSort(searchResults, "ibu", false);
    sortedByIbuAsc.reduce(function(accumulator, currentValue) {
      expect(accumulator.ibu < currentValue.ibu).toBeTruthy();
      return currentValue;
    });

    const sortedByAbvAsc = wrapper.vm.customSort(searchResults, "abv", false);
    sortedByAbvAsc.reduce(function(accumulator, currentValue) {
      expect(accumulator.abv < currentValue.abv).toBeTruthy();
      return currentValue;
    });

    const sortedByAbvDesc = wrapper.vm.customSort(searchResults, "abv", true);
    sortedByAbvDesc.reduce(function(accumulator, currentValue) {
      expect(accumulator.abv > currentValue.abv).toBeTruthy();
      return currentValue;
    });
  });

  it("compares dates with compareDates", () => {
    const dateA = "09/2007";
    const dateB = "10/2007";
    expect(wrapper.vm.compareDates(dateA, dateB)).toEqual(-1);
    const dateC = "08/2007";
    expect(wrapper.vm.compareDates(dateA, dateC)).toEqual(1);
  });

  it("handles empty headers array", () => {
    wrapper = shallowMount(SearchResults);
    expect(wrapper.vm.headers.length).toEqual(0);
  });

  it("SearchResults has same HTML structure", () => {
    const renderer = createRenderer();
    renderer.renderToString(wrapper.vm, (err, str) => {
      if (err) throw new Error(err);
      expect(str).toMatchSnapshot();
    });
  });
});
