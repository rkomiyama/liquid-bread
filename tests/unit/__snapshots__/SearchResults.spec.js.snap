// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchResults.vue SearchResults has same HTML structure 1`] = `
<v-data-table-stub noDataText="$vuetify.noDataText" noResultsText="$vuetify.dataIterator.noResultsText" nextIcon="$vuetify.icons.next" prevIcon="$vuetify.icons.prev" rowsPerPageItems="" rowsPerPageText="$vuetify.dataTable.rowsPerPageText" search="" filter="function _default(val, search) {
                return val != null &amp;&amp; typeof val !== 'boolean' &amp;&amp; val.toString().toLowerCase().indexOf(search) !== -1;
            }" customFilter="function _default(items, search, filter, headers) {
                search = search.toString().toLowerCase();
                if (search.trim() === '') return items;
                var props = headers.map(function (h) {
                    return h.value;
                });
                return items.filter(function (item) {
                    return props.some(function (prop) {
                        return filter(Object(_util_helpers__WEBPACK_IMPORTED_MODULE_7__[&quot;getObjectValueByPath&quot;])(item, prop, item[prop]), search);
                    });
                });
            }" customSort="function () { [native code] }" value="" items="[object Object],[object Object],[object Object],[object Object],[object Object]" itemKey="id" sortIcon="$vuetify.icons.sort" headers="[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]" headerText="text">
  <v-progress-linear-stub color="blue" active="true" bufferValue="100" height="7" indeterminate="indeterminate" value="0"></v-progress-linear-stub>
</v-data-table-stub>
`;
